File: src/mem/slicc/ast/PeekStatementAST.cc

/*
 * Copyright (c) 1999-2008 Mark D. Hill and David A. Wood
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer;
 * redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution;
 * neither the name of the copyright holders nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * PeekStatementAST.C
 *
 * Description: See PeekStatementAST.h
 *
 * $Id$
 *
 */

#include "PeekStatementAST.hh"
#include "SymbolTable.hh"
#include "StatementListAST.hh"
#include "TypeAST.hh"
#include "VarExprAST.hh"

PeekStatementAST::PeekStatementAST(VarExprAST* queue_name_ptr,
                                   TypeAST* type_ptr,
                                   StatementListAST* statementlist_ptr,
                                   string method)
  : StatementAST()
{
  m_queue_name_ptr = queue_name_ptr;
  m_type_ptr = type_ptr;
  m_statementlist_ptr = statementlist_ptr;
  m_method = method;
}

PeekStatementAST::~PeekStatementAST()
{
  delete m_queue_name_ptr;
  delete m_type_ptr;
  delete m_statementlist_ptr;
}

void PeekStatementAST::generate(string& code, Type* return_type_ptr) const
{
  code += indent_str() + "{\n";  // Start scope
  inc_indent();
  g_sym_table.pushFrame();

  Type* msg_type_ptr = m_type_ptr->lookupType();

  // Add new local var to symbol table
  g_sym_table.newSym(new Var("in_msg", getLocation(), msg_type_ptr, "(*in_msg_ptr)", getPairs()));

  // Check the queue type
  m_queue_name_ptr->assertType("InPort");

  // Declare the new "in_msg_ptr" variable
  code += indent_str() + "const " + msg_type_ptr->cIdent() + "* in_msg_ptr;\n";  // Declare message
  //  code += indent_str() + "in_msg_ptr = static_cast<const ";
  code += indent_str() + "in_msg_ptr = dynamic_cast<const ";
  code += msg_type_ptr->cIdent() + "*>(";
  code += "(" + m_queue_name_ptr->getVar()->getCode() + ")";
  code += ".";
  code += m_method;
  code += "());\n";

  code += indent_str() + "assert(in_msg_ptr != NULL);\n";        // Check the cast result

  if(CHECK_INVALID_RESOURCE_STALLS) {
    // Declare the "in_buffer_rank" variable
    code += indent_str() + "int in_buffer_rank = ";  // Declare message
    code += "(" + m_queue_name_ptr->getVar()->getCode() + ")";
    code += ".getPriority();\n";
  }

  m_statementlist_ptr->generate(code, return_type_ptr);                // The other statements
  dec_indent();
  g_sym_table.popFrame();
  code += indent_str() + "}\n";  // End scope
}

void PeekStatementAST::findResources(Map<Var*, string>& resource_list) const
{
  m_statementlist_ptr->findResources(resource_list);
}

void PeekStatementAST::print(ostream& out) const
{
  out << "[PeekStatementAST: " << m_method
      << " queue_name: " << *m_queue_name_ptr
      << " type: " << m_type_ptr->toString()
      << " " << *m_statementlist_ptr
      << "]";
}

File: src/mem/slicc/ast/PeekStatementAST.py
# Copyright (c) 2013 Advanced Micro Devices, Inc.
# Copyright (c) 1999-2008 Mark D. Hill and David A. Wood
# Copyright (c) 2009 The Hewlett-Packard Development Company
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met: redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer;
# redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution;
# neither the name of the copyright holders nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from slicc.ast.StatementAST import StatementAST
from slicc.symbols import Var


class PeekStatementAST(StatementAST):
    def __init__(self, slicc, queue_name, type_ast, pairs, statements, method):
        super().__init__(slicc, pairs)

        self.queue_name = queue_name
        self.type_ast = type_ast
        self.statements = statements
        self.method = method

    def __repr__(self):
        return (
            "[PeekStatementAST: {!r} queue_name: {!r} type: {!r} {!r}]".format(
                self.method,
                self.queue_name,
                self.type_ast,
                self.statements,
            )
        )

    def generate(self, code, return_type, **kwargs):
        self.symtab.pushFrame()

        msg_type = self.type_ast.type

        # Add new local var to symbol table
        var = Var(
            self.symtab,
            "in_msg",
            self.location,
            msg_type,
            "(*in_msg_ptr)",
            self.pairs,
        )
        self.symtab.newSymbol(var)

        # Check the queue type
        self.queue_name.assertType("InPort")

        # Declare the new "in_msg_ptr" variable
        mtid = msg_type.c_ident
        qcode = self.queue_name.var.code
        code(
            """
{
    // Declare message
    [[maybe_unused]] const $mtid* in_msg_ptr;
    in_msg_ptr = dynamic_cast<const $mtid *>(($qcode).${{self.method}}());
    if (in_msg_ptr == NULL) {
        // If the cast fails, this is the wrong inport (wrong message type).
        // Throw an exception, and the caller will decide to either try a
        // different inport or punt.
        throw RejectException();
    }
"""
        )

        if "block_on" in self.pairs:
            address_field = self.pairs["block_on"]
            code(
                """
    if (m_is_blocking &&
        (m_block_map.count(in_msg_ptr->m_$address_field) == 1) &&
        (m_block_map[in_msg_ptr->m_$address_field] != &$qcode)) {
            $qcode.delayHead(clockEdge(), cyclesToTicks(Cycles(1)),
            m_ruby_system->getRandomization(), m_ruby_system->getWarmupEnabled());
            continue;
    }
            """
            )

        if "wake_up" in self.pairs:
            address_field = self.pairs["wake_up"]
            code(
                """
    if (m_waiting_buffers.count(in_msg_ptr->m_$address_field) > 0) {
        wakeUpBuffers(in_msg_ptr->m_$address_field);
    }
            """
            )

        # The other statements
        self.statements.generate(code, return_type, **kwargs)
        self.symtab.popFrame()
        code("}")

    def findResources(self, resources):
        self.statements.findResources(resources)
